### PR TITLE
123

### PR DESCRIPTION
**Pull request by EliosPride**

### CODE REVIEW

File: src/main/java/core/basesyntax/entity/FruitTransaction.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
+package core.basesyntax.entity;
+
+import com.opencsv.bean.CsvBindByName;
+import com.opencsv.bean.CsvCustomBindByName;
+import core.basesyntax.converter.ContentTypeEnumConverter;
+
+public class FruitTransaction {
+    @CsvCustomBindByName(converter = ContentTypeEnumConverter.class)
+    private Operation type;
+    @CsvBindByName(column = "fruit")
```
Comment (al-pavlenko) [inline src/main/java/core/basesyntax/entity/FruitTransaction.java]: column param is redundant

File: src/main/java/core/basesyntax/entity/Operation.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    RETURN("r");
+
+    private final String operation;
+
+    Operation(String operation) {
+        this.operation = operation;
+    }
+
+    public static Operation fromString(String operation) {
+        for (Operation b : Operation.values()) {
```
Comment (al-pavlenko) [inline src/main/java/core/basesyntax/entity/Operation.java]: rename to 'element'

File: src/main/java/core/basesyntax/Main.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+import core.basesyntax.converter.ConverterCSVToObject;
+import core.basesyntax.converter.ConverterObjectToCSV;
+import core.basesyntax.service.FruitOperationGenerator;
+
+import java.io.IOException;
+
+public class Main {
+    public static void main(String[] args) throws IOException, CsvRequiredFieldEmptyException, CsvDataTypeMismatchException {
+        String inputFileName = "src/main/resources/fruitShop.csv";
+        ConverterCSVToObject converterCSVToObject = new ConverterCSVToObject();
```
Comment (al-pavlenko) [inline src/main/java/core/basesyntax/Main.java]: move methods from ConverterCSVToObject and ConverterObjectToCSV to common class FileUtils (in package 'util') and make them static.

File: src/main/java/core/basesyntax/Main.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+import com.opencsv.exceptions.CsvRequiredFieldEmptyException;
+import core.basesyntax.converter.ConverterCSVToObject;
+import core.basesyntax.converter.ConverterObjectToCSV;
+import core.basesyntax.service.FruitOperationGenerator;
+
+import java.io.IOException;
+
+public class Main {
+    public static void main(String[] args) throws IOException, CsvRequiredFieldEmptyException, CsvDataTypeMismatchException {
+        String inputFileName = "src/main/resources/fruitShop.csv";
```
Comment (al-pavlenko) [inline src/main/java/core/basesyntax/Main.java]: move to constant in FileUtils

File: src/main/java/core/basesyntax/Main.java [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+        ConverterCSVToObject converterCSVToObject = new ConverterCSVToObject();
+        converterCSVToObject.reader(inputFileName);
```
Comment (al-pavlenko) [inline src/main/java/core/basesyntax/Main.java]: rename to 'readFruitTransactions'

File: src/main/java/core/basesyntax/Main.java [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+        FruitOperationGenerator fruitOperationGenerator = new FruitOperationGenerator();
+        fruitOperationGenerator.fruitRestReport();
+
+        ConverterObjectToCSV converterObjectToCSV = new ConverterObjectToCSV();
+        converterObjectToCSV.writer();
```
Comment (al-pavlenko) [inline src/main/java/core/basesyntax/Main.java]: rename to 'generateFruitReport'

File: src/main/java/core/basesyntax/converter/ContentTypeEnumConverter.java [hunk 7]
```diff
@@ -0,0 +1,11 @@
+package core.basesyntax.converter;
+
+import com.opencsv.bean.AbstractBeanField;
+import core.basesyntax.entity.Operation;
+
+public class ContentTypeEnumConverter extends AbstractBeanField<Operation> {
```
Comment (al-pavlenko) [inline src/main/java/core/basesyntax/converter/ContentTypeEnumConverter.java]: rename 'OperationConverter'

File: src/main/java/core/basesyntax/converter/ConverterCSVToObject.java [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+
+import java.io.FileNotFoundException;
+import java.io.FileReader;
+import java.util.List;
+
+public class ConverterCSVToObject {
+
+    public List<FruitTransaction> reader(String inputFileName) throws FileNotFoundException {
+        List<FruitTransaction> fruitTransactionList;
+        fruitTransactionList = new CsvToBeanBuilder<FruitTransaction>(new FileReader(inputFileName))
```
Comment (al-pavlenko) [inline src/main/java/core/basesyntax/converter/ConverterCSVToObject.java]: do return here inline, no need to create variable

File: src/main/java/core/basesyntax/converter/ConverterCSVToObject.java [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+                .withType(FruitTransaction.class)
+                .build()
+                .parse();
+        fruitTransactionList.forEach(System.out::println);
```
Comment (al-pavlenko) [inline src/main/java/core/basesyntax/converter/ConverterCSVToObject.java]: remove this

File: src/main/java/core/basesyntax/converter/ConverterObjectToCSV.java [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+import core.basesyntax.db.FruitStorage;
+
+import java.io.FileWriter;
+import java.io.IOException;
+import java.io.Writer;
+
+public class ConverterObjectToCSV {
+
+    public void writer() throws IOException, CsvRequiredFieldEmptyException, CsvDataTypeMismatchException {
+        String outputFile = "src/main/resources/fruitReport.csv";
```
Comment (al-pavlenko) [inline src/main/java/core/basesyntax/converter/ConverterObjectToCSV.java]: move to constant

File: src/main/java/core/basesyntax/converter/ConverterCSVToObject.java [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+import com.opencsv.bean.CsvToBeanBuilder;
+import core.basesyntax.entity.FruitTransaction;
+
+import java.io.FileNotFoundException;
+import java.io.FileReader;
+import java.util.List;
+
+public class ConverterCSVToObject {
+
+    public List<FruitTransaction> reader(String inputFileName) throws FileNotFoundException {
```
Comment (al-pavlenko) [inline src/main/java/core/basesyntax/converter/ConverterCSVToObject.java]: wrap with try with resources

File: src/main/java/core/basesyntax/converter/ConverterObjectToCSV.java [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+        Writer writer = new FileWriter(outputFile);
+        StatefulBeanToCsv<Fruit> sbc = new StatefulBeanToCsvBuilder<Fruit>(writer)
+                .withSeparator(CSVWriter.DEFAULT_SEPARATOR)
+                .build();
+        sbc.write(FruitStorage.FRUIT_STORAGE);
+        writer.close();
```
Comment (al-pavlenko) [inline src/main/java/core/basesyntax/converter/ConverterObjectToCSV.java]: use try with resources

File: src/main/java/core/basesyntax/service/FruitOperationGenerator.java [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+
+import java.io.FileNotFoundException;
+import java.util.List;
+
+public class FruitOperationGenerator {
+
+    public void fruitRestReport() throws FileNotFoundException {
+        String inputFile = "src/main/resources/fruitShop.csv";
+        ConverterCSVToObject converterCSVToObject = new ConverterCSVToObject();
+        List<FruitTransaction> fruitTransactionList = converterCSVToObject.reader(inputFile);
```
Comment (al-pavlenko) [inline src/main/java/core/basesyntax/service/FruitOperationGenerator.java]: you have already read it in Main, call forEach in Main and just pass each fruit here as argument

File: src/main/java/core/basesyntax/service/FruitOperationGenerator.java [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+import core.basesyntax.db.FruitStorage;
+import core.basesyntax.entity.FruitTransaction;
+import core.basesyntax.converter.ConverterCSVToObject;
+
+import java.io.FileNotFoundException;
+import java.util.List;
+
+public class FruitOperationGenerator {
+
+    public void fruitRestReport() throws FileNotFoundException {
```
Comment (al-pavlenko) [inline src/main/java/core/basesyntax/service/FruitOperationGenerator.java]: rename to 'process'

File: src/main/java/core/basesyntax/service/FruitOperationGenerator.java [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+
+import core.basesyntax.entity.Fruit;
+import core.basesyntax.db.FruitStorage;
+import core.basesyntax.entity.FruitTransaction;
+import core.basesyntax.converter.ConverterCSVToObject;
+
+import java.io.FileNotFoundException;
+import java.util.List;
+
+public class FruitOperationGenerator {
```
Comment (al-pavlenko) [inline src/main/java/core/basesyntax/service/FruitOperationGenerator.java]: rename to 'FruitProcessor'

File: src/main/java/core/basesyntax/service/FruitOperationGenerator.java [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+        List<FruitTransaction> fruitTransactionList = converterCSVToObject.reader(inputFile);
+        FruitOperationImpl fruitOperation = new FruitOperationImpl();
+        fruitTransactionList.forEach(x -> {
+            switch (x.getType()) {
+                case BALANCE -> {
+                    Fruit fruit = new Fruit(x.getFruit(), x.getQuantity());
+                    fruitOperation.transaction(fruit);
+                }
+                case SUPPLY, RETURN -> {
+                    Fruit fruit = new Fruit(x.getFruit(), x.getQuantity());
```
Comment (al-pavlenko) [inline src/main/java/core/basesyntax/service/FruitOperationGenerator.java]: no need to create Fruit object, take values from var 'x'

File: src/main/java/core/basesyntax/service/FruitOperationGenerator.java [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+                    FruitStorage.FRUIT_STORAGE.stream()
+                            .filter(a -> a.getName().equals(fruit.getName()))
+                            .findFirst()
+                            .ifPresent(a -> a.setAmount(a.getAmount() + fruit.getAmount()));
+                }
+                case PURCHASE -> {
+                    Fruit fruit = new Fruit(x.getFruit(), x.getQuantity());
```
Comment (al-pavlenko) [inline src/main/java/core/basesyntax/service/FruitOperationGenerator.java]: no need to create Fruit object, take values from var 'x'

File: src/main/java/core/basesyntax/service/FruitOperationGenerator.java [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+        fruitTransactionList.forEach(x -> {
+            switch (x.getType()) {
+                case BALANCE -> {
+                    Fruit fruit = new Fruit(x.getFruit(), x.getQuantity());
+                    fruitOperation.transaction(fruit);
+                }
+                case SUPPLY, RETURN -> {
+                    Fruit fruit = new Fruit(x.getFruit(), x.getQuantity());
+                    FruitStorage.FRUIT_STORAGE.stream()
+                            .filter(a -> a.getName().equals(fruit.getName()))
```
Comment (al-pavlenko) [inline src/main/java/core/basesyntax/service/FruitOperationGenerator.java]: create static method 'getByName' in FruitStorage that will return Optional<Fruit>

File: src/main/java/core/basesyntax/service/FruitOperationGenerator.java [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+                            .findFirst()
+                            .ifPresent(a -> a.setAmount(a.getAmount() + fruit.getAmount()));
+                }
+                case PURCHASE -> {
+                    Fruit fruit = new Fruit(x.getFruit(), x.getQuantity());
+                    FruitStorage.FRUIT_STORAGE.stream()
+                            .filter(q -> q.getName().equals(fruit.getName()))
```
Comment (al-pavlenko) [inline src/main/java/core/basesyntax/service/FruitOperationGenerator.java]: use getByName from previous comment )

File: src/main/java/core/basesyntax/service/FruitOperationImpl.java [hunk 20]
```diff
@@ -0,0 +1,11 @@
+package core.basesyntax.service;
+
+import core.basesyntax.entity.Fruit;
+import core.basesyntax.db.FruitStorage;
+
+public class FruitOperationImpl implements FruitOperation {
+    @Override
+    public void transaction(Fruit fruit) {
```
Comment (al-pavlenko) [inline src/main/java/core/basesyntax/service/FruitOperationImpl.java]: create method 'add' in FruitStorage and use it for save fruits in all places. This class can be removed

File: src/main/java/core/basesyntax/converter/ConverterObjectToCSV.java [hunk 21]
```diff
@@ -0,0 +1,26 @@
...
+
+public class ConverterObjectToCSV {
+
+    public void writer() throws IOException, CsvRequiredFieldEmptyException, CsvDataTypeMismatchException {
+        String outputFile = "src/main/resources/fruitReport.csv";
+        Writer writer = new FileWriter(outputFile);
+        StatefulBeanToCsv<Fruit> sbc = new StatefulBeanToCsvBuilder<Fruit>(writer)
+                .withSeparator(CSVWriter.DEFAULT_SEPARATOR)
+                .build();
+        sbc.write(FruitStorage.FRUIT_STORAGE);
```
Comment (al-pavlenko) [inline src/main/java/core/basesyntax/converter/ConverterObjectToCSV.java]: create separate class for report like 'FruitReport', convert here fruits to List<FruitReport> and write it, not Fruit

File: src/main/java/core/basesyntax/entity/Operation.java (hunk 2 above)
Comment (EliosPride) [reply src/main/java/core/basesyntax/entity/Operation.java]: done

File: src/main/java/core/basesyntax/Main.java (hunk 3 above)
Comment (EliosPride) [reply src/main/java/core/basesyntax/Main.java]: done

File: src/main/java/core/basesyntax/Main.java (hunk 4 above)
Comment (EliosPride) [reply src/main/java/core/basesyntax/Main.java]: done

File: src/main/java/core/basesyntax/Main.java (hunk 5 above)
Comment (EliosPride) [reply src/main/java/core/basesyntax/Main.java]: @done

File: src/main/java/core/basesyntax/Main.java (hunk 6 above)
Comment (EliosPride) [reply src/main/java/core/basesyntax/Main.java]: @al-pavlenko done

File: src/main/java/core/basesyntax/converter/ContentTypeEnumConverter.java (hunk 7 above)
Comment (EliosPride) [reply src/main/java/core/basesyntax/converter/ContentTypeEnumConverter.java]: @al-pavlenko done

File: src/main/java/core/basesyntax/converter/ConverterCSVToObject.java (hunk 8 above)
Comment (EliosPride) [reply src/main/java/core/basesyntax/converter/ConverterCSVToObject.java]: @al-pavlenko done

File: src/main/java/core/basesyntax/converter/ConverterCSVToObject.java (hunk 9 above)
Comment (EliosPride) [reply src/main/java/core/basesyntax/converter/ConverterCSVToObject.java]: @al-pavlenko done

File: src/main/java/core/basesyntax/converter/ConverterObjectToCSV.java (hunk 10 above)
Comment (EliosPride) [reply src/main/java/core/basesyntax/converter/ConverterObjectToCSV.java]: @al-pavlenko done

File: src/main/java/core/basesyntax/service/FruitOperationGenerator.java (hunk 13 above)
Comment (EliosPride) [reply src/main/java/core/basesyntax/service/FruitOperationGenerator.java]: @al-pavlenko done

File: src/main/java/core/basesyntax/service/FruitOperationGenerator.java (hunk 14 above)
Comment (EliosPride) [reply src/main/java/core/basesyntax/service/FruitOperationGenerator.java]: @al-pavlenko done

File: src/main/java/core/basesyntax/service/FruitOperationGenerator.java (hunk 15 above)
Comment (EliosPride) [reply src/main/java/core/basesyntax/service/FruitOperationGenerator.java]: @al-pavlenko done

File: src/main/java/core/basesyntax/service/FruitOperationGenerator.java (hunk 16 above)
Comment (EliosPride) [reply src/main/java/core/basesyntax/service/FruitOperationGenerator.java]: @al-pavlenko done

File: src/main/java/core/basesyntax/service/FruitOperationGenerator.java (hunk 17 above)
Comment (EliosPride) [reply src/main/java/core/basesyntax/service/FruitOperationGenerator.java]: @al-pavlenko done

File: src/main/java/core/basesyntax/Main.java (hunk 4 above)
Comment (al-pavlenko) [reply src/main/java/core/basesyntax/Main.java]: @EliosPride What have you done?

File: src/main/java/core/basesyntax/Main.java [hunk 22]
```diff
@@ -0,0 +1,20 @@
...
+import com.opencsv.exceptions.CsvRequiredFieldEmptyException;
+import core.basesyntax.service.FruitProcessor;
+import core.basesyntax.util.FileUtil;
+
+import java.io.IOException;
+
+public class Main {
+    public static void main(String[] args) throws IOException, CsvRequiredFieldEmptyException, CsvDataTypeMismatchException {
+        String inputFileName = "src/main/resources/fruitShop.csv";
+        FileUtil.readFruitTransactions(inputFileName);
```
Comment (al-pavlenko) [inline src/main/java/core/basesyntax/Main.java]: What are you doing here?

File: src/main/java/core/basesyntax/entity/Operation.java (hunk 2 above)
Comment (al-pavlenko) [reply src/main/java/core/basesyntax/entity/Operation.java]: @EliosPride and again, what have you done?

File: src/main/java/core/basesyntax/service/FruitProcessor.java [hunk 23]
```diff
@@ -0,0 +1,25 @@
...
+import core.basesyntax.util.FileUtil;
+
+import java.util.List;
+
+public class FruitProcessor {
+
+    public void process() {
+        String inputFile = "src/main/resources/fruitShop.csv";
+        List<FruitTransaction> fruitTransactionList = FileUtil.readFruitTransactions(inputFile);
+        fruitTransactionList.forEach(x -> {
```
Comment (al-pavlenko) [inline src/main/java/core/basesyntax/service/FruitProcessor.java]: this is processor, so it has just to process incoming fruits, don't read here from file

File: src/main/java/core/basesyntax/util/FileUtil.java [hunk 24]
```diff
@@ -0,0 +1,39 @@
...
+            System.out.println("error occurred during reading file");
+        }
+        return Collections.emptyList();
+    }
+
+    public static void generateFruitReport() {
+        try (Writer writer = new FileWriter(OUTPUT_FILE); CSVWriter csvWriter = new CSVWriter(writer)) {
+            for (String[] list : FruitReport.fruitReport()) {
+                csvWriter.writeNext(list);
+            }
```
Comment (al-pavlenko) [inline src/main/java/core/basesyntax/util/FileUtil.java]: Why have you changed write logic? Previous was better

File: src/main/java/core/basesyntax/service/FruitReport.java [hunk 25]
```diff
@@ -0,0 +1,15 @@
...
+
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+public class FruitReport {
+    public static List<String[]> fruitReport() {
+        return FruitStorage.FRUIT_STORAGE.stream()
+                .map(fruit -> new String[]{Objects.toString(fruit, null)})
+                .collect(Collectors.toList());
```
Comment (al-pavlenko) [inline src/main/java/core/basesyntax/service/FruitReport.java]: you need to create class with fields you want to export and map them like in FruitTransaction.

File: src/main/resources/fruitReport.csv [hunk 26]
```diff
@@ -0,0 +1,2 @@
+"core.basesyntax.entity.Fruit@12bb4df8"
+"core.basesyntax.entity.Fruit@4cc77c2e"
```
Comment (al-pavlenko) [inline src/main/resources/fruitReport.csv]: is this a report?

File: src/main/java/core/basesyntax/Main.java [hunk 27]
```diff
@@ -0,0 +1,20 @@
...
+
+import com.opencsv.exceptions.CsvDataTypeMismatchException;
+import com.opencsv.exceptions.CsvRequiredFieldEmptyException;
+import core.basesyntax.service.FruitProcessor;
+import core.basesyntax.util.FileUtil;
+
+import java.io.IOException;
+
+public class Main {
+    public static void main(String[] args) throws IOException, CsvRequiredFieldEmptyException, CsvDataTypeMismatchException {
```
Comment (al-pavlenko) [inline src/main/java/core/basesyntax/Main.java]: dont throw exceptions from main

**Pull request merged**